In [1]:
# Imports
import pandas as pd
import numpy as np

### WIP

Ideas:
    1.add introduction
    2.add methods (?)
    3.provide brief introduction to the data
    4.


### WIP


### Context
"Predict behavor to retain customers. You can analyze all relevant customer data and develop focused customer retention programs."

### Content
Each row represents a customer, each column contains customer’s attributes described on the column Metadata.

The data set includes information about:

Customers who left within the last month – the column is called Churn
Services that each customer has signed up for – phone, multiple lines, internet, online security, online backup, device protection, tech support, and streaming TV and movies
Customer account information – how long they’ve been a customer, contract, payment method, paperless billing, monthly charges, and total charges
Demographic info about customers – gender, age range, and if they have partners and dependents

In [2]:
# Load & print initial info
df = pd.read_csv('data/WA_Fn-UseC_-Telco-Customer-Churn.csv')
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [3]:
# Initial observations
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


### WIP add short data description

It's possible to see our dataset has 21 columns/features and 7,043 observations. Most of the features are made of string data type, including 'TotalCharges', which appear to be made of numbers and not letters, and we have 3 columns made of numbers, either as integers or float64. It is possible to see that we have a substantial number of features made of categorical data, which will increase the number of columns after treatment for use with statistical models.



In [4]:
len(set(df['customerID']))



7043

In [5]:
# Columns with string data type
string_cols = [
    'customerID',
    'gender',
    'Partner',
    'Dependents',
    'PhoneService',
    'MultipleLines',
    'InternetService',
    'OnlineSecurity',
    'OnlineBackup',
    'DeviceProtection',
    'TechSupport',
    'StreamingTV',
    'StreamingMovies',
    'Contract',
    'PaperlessBilling',
    'PaymentMethod',
    'TotalCharges',
    'Churn' 
]

# Columns with numerical data type
num_cols = [
    'SeniorCitizen',
    'tenure',
    'MonthlyCharges',
]

# CHeck possible values from string data types
for col in string_cols:
    unique_values = set(df[col])
    print(f"{col}\n{len(col)*'-'}")
    print(set(df[col]), '\n')

customerID
----------
{'9555-SAHUZ', '9488-FVZCC', '4114-QMKVN', '3877-JRJIP', '0823-HSCDJ', '8229-BUJHX', '2516-XSJKX', '6248-TKCQV', '9717-QEBGU', '2929-QNSRW', '1845-CSBRZ', '5498-IBWPI', '4831-EOBFE', '2480-EJWYP', '0471-ARVMX', '3566-HJGPK', '8084-OIVBS', '2690-DVRVK', '5327-XOKKY', '7912-SYRQT', '2226-ICFDO', '9168-INPSZ', '4536-PLEQY', '6496-JDSSB', '6789-HJBWG', '5067-DGXLL', '7319-ZNRTR', '5406-KGRMX', '0214-JHPFW', '3096-YXENJ', '3753-TSEMP', '1536-HBSWP', '6351-SCJKT', '0750-EKNGL', '2808-CHTDM', '0678-RLHVP', '8180-AKMJV', '9921-ZVRHG', '3317-HRTNN', '3091-FYHKI', '5539-TMZLF', '1131-ALZWV', '1867-TJHTS', '1951-IEYXM', '5567-WSELE', '0430-IHCDJ', '8679-LZBMD', '6575-SUVOI', '3779-OSWCF', '8292-ITGYJ', '6156-UZDLF', '3451-VAWLI', '3696-DFHHB', '5968-XQIVE', '5848-FHRFC', '5906-CVLHP', '8713-IGZSO', '3863-IUBJR', '1830-GGFNM', '1293-HHSHJ', '8687-BAFGU', '5343-SGUBI', '8089-UZWLX', '2379-ENZGV', '9239-ZBZZV', '3678-MNGZX', '9254-RBFON', '6075-SLNIL', '6718-BDGHG', '3445-HXXGF

In [6]:
for col in num_cols:
    print(f"{col}\n{len(col)*'-'}")
    print(set(df[col]), '\n')

SeniorCitizen
-------------
{0, 1} 

tenure
------
{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72} 

MonthlyCharges
--------------
{18.95, 19.8, 20.65, 20.15, 20.2, 20.75, 24.95, 20.4, 24.25, 25.9, 19.95, 29.75, 29.85, 30.2, 24.3, 33.75, 30.15, 31.05, 35.45, 29.95, 30.4, 39.65, 35.9, 40.2, 42.3, 42.1, 44.35, 45.25, 45.55, 45.65, 45.3, 49.95, 49.55, 49.25, 49.05, 53.85, 54.4, 55.2, 56.95, 56.15, 55.3, 59.9, 59.6, 55.65, 54.65, 62.15, 64.7, 64.15, 66.15, 64.5, 66.85, 69.7, 70.7, 69.2, 69.5, 70.45, 74.8, 74.85, 76.2, 75.3, 78.9, 79.85, 80.65, 79.75, 79.2, 74.75, 84.5, 79.35, 82.05, 84.0, 85.2, 89.1, 90.05, 90.25, 89.9, 91.65, 94.4, 95.5, 96.35, 97.85, 95.45, 99.65, 100.35, 99.35, 102.45, 102.7, 103.8, 103.4, 104.15, 104.4, 104.6, 105.35, 108.0, 109.2, 103.7,

### wip

we see that the dataset has a significant number of categorical features, with most being 2/binary (e.g. 'yes'/'no') values, which is the case for the goal-feature, or engoenous variable, 'Churn', but also for a numerical 0/1 'SeniorCitizen' feature.

In [7]:
# Adapt 'TotalCharges' data type
df_adapted = df.copy()
df_adapted['TotalCharges'] = pd.to_numeric(df_adapted['TotalCharges'], errors='coerce')

# Check for missing values
count_missing_values = df_adapted['TotalCharges'].isna().sum().item()
print(f'Missing TotalCharges: {count_missing_values}')

Missing TotalCharges: 11


In [ ]:
# Check for additional information
df.loc[df_adapted[df_adapted['TotalCharges'].isna()].index][['TotalCharges', 'Churn']]

,TotalCharges,Churn
488,,No
753,,No
936,,No
1082,,No
1340,,No
3331,,No
3826,,No
4380,,No
5218,,No
6670,,No


In [31]:
df['Churn'].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

Checking the values turned into null by the function pd.to_numeric() we see that they are all empty strings and there is no transformation we could have done to them. We are left with 2 options: 1. droping them; 2. predicting them using a statistical method. Since we have only 11 observations out of 7,043 and they are all related to no churn, we'll choose to drop the observations.